runtime/common/src/claims.rs

```diff
@@ -301,38 +429,105 @@ impl<T: Trait> sp_runtime::traits::ValidateUnsigned for Module<T> {
-						ValidityError::SignerHasNoClaim.into(),
-					).into());
+				Self::eth_recover(&ethereum_signature, &data, &[][..])
+			}
+			Call::claim_attest(account, ethereum_signature, statement) => {
+				let data = account.using_encoded(to_ascii_hex);
+				let maybe_signer = Self::eth_recover(&ethereum_signature, &data, statement.to_text());
+				if let Some(ref signer) = maybe_signer {
+					if let Some(s) = Signing::get(signer) {
+						let e = InvalidTransaction::Custom(ValidityError::InvalidStatement.into());
+						ensure!(&s == statement, e);
+					}
 				}
+				maybe_signer
+			}
+			_ => return Err(InvalidTransaction::Call.into()),
+		};
+
+		let signer = maybe_signer
+			.ok_or(InvalidTransaction::Custom(ValidityError::InvalidEthereumSignature.into()))?;
+
+		let e = InvalidTransaction::Custom(ValidityError::SignerHasNoClaim.into());
+		ensure!(<Claims<T>>::contains_key(&signer), e);
+
+		Ok(ValidTransaction {
+			priority: PRIORITY,
+			requires: vec![],
+			provides: vec![("claims", signer).encode()],
+			longevity: TransactionLongevity::max_value(),
+			propagate: true,
+		})
+	}
+}
 
-				Ok(ValidTransaction {
-					priority: PRIORITY,
-					requires: vec![],
-					provides: vec![("claims", signer).encode()],
-					longevity: TransactionLongevity::max_value(),
-					propagate: true,
-				})
+/// Validate `attest` calls prior to execution. Needed to avoid a DoS attack since they are
+/// otherwise free to place on chain.
+#[derive(Encode, Decode, Clone, Eq, PartialEq)]
+pub struct PrevalidateAttests<T: Trait + Send + Sync>(sp_std::marker::PhantomData<T>) where
+	<T as system::Trait>::Call: IsSubType<Module<T>, T>;
+
+impl<T: Trait + Send + Sync> Debug for PrevalidateAttests<T> where
+	<T as system::Trait>::Call: IsSubType<Module<T>, T>
+{
+	#[cfg(feature = "std")]
+	fn fmt(&self, f: &mut sp_std::fmt::Formatter) -> sp_std::fmt::Result {
+		write!(f, "PrevalidateAttests")
+	}
+
+	#[cfg(not(feature = "std"))]
+	fn fmt(&self, _: &mut sp_std::fmt::Formatter) -> sp_std::fmt::Result {
+		Ok(())
+	}
+}
+
+impl<T: Trait + Send + Sync> PrevalidateAttests<T> where
+	<T as system::Trait>::Call: IsSubType<Module<T>, T>
+{
+	/// Create new `SignedExtension` to check runtime version.
+	pub fn new() -> Self {
+		Self(sp_std::marker::PhantomData)
+	}
+}
+
+impl<T: Trait + Send + Sync> SignedExtension for PrevalidateAttests<T> where
+	<T as system::Trait>::Call: IsSubType<Module<T>, T>
+{
+	type AccountId = T::AccountId;
+	type Call = <T as system::Trait>::Call;
+	type AdditionalSigned = ();
+	type Pre = ();
+	const IDENTIFIER: &'static str = "PrevalidateAttests";
+
+	fn additional_signed(&self) -> Result<Self::AdditionalSigned, TransactionValidityError> {
+		Ok(())
+	}
+
+	fn validate(
+		&self,
+		who: &Self::AccountId,
+		call: &Self::Call,
+		_info: &DispatchInfoOf<Self::Call>,
+		_len: usize,
+	) -> TransactionValidity {
+		if let Some(local_call) = call.is_sub_type() {
+			if let Call::attest(attested_statement) = local_call {
```

Comment:
What would happen if the raw transaction of the giant size? It won't pass this validation, sure, but it would need to be first deserialized to get here.

Reply:
I think the whole validation pipeline can not do anything from stopping a single node from griefing. Instead, the only thing we can provide is safety to the whole network that this node (any any other node that does this validation), will not forward the transaction to their peers.

Reply:
indeed. this comment holds for all transaction types.